lib/src/time_util.rs

```diff
@@ -0,0 +1,191 @@
+// Copyright 2024 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Provides support for parsing and matching date ranges.
+
+use chrono::{DateTime, FixedOffset, Local, TimeZone, Utc};
+use chrono_english::{parse_date_string, DateError, Dialect};
+use thiserror::Error;
+
+use crate::backend::{MillisSinceEpoch, Timestamp};
+
+/// Context needed to create a DatePattern during revset evaluation.
+#[derive(Clone, Debug)]
+pub enum DatePatternContext {
+    /// Interpret date patterns using the local machine's time zone
+    Local(DateTime<Local>),
+    /// Interpret date patterns using UTC
+    Utc(DateTime<Utc>),
+    /// Interpret date patterns using any FixedOffset time zone
+    Fixed(DateTime<FixedOffset>),
+}
+
+impl DatePatternContext {
+    /// Parses a DatePattern from the given string and kind.
+    pub fn parse_relative(
+        &self,
+        s: &str,
+        kind: &str,
+    ) -> Result<DatePattern, DatePatternParseError> {
+        match *self {
+            DatePatternContext::Local(dt) => DatePattern::from_str_kind(s, kind, dt.to_owned()),
+            DatePatternContext::Utc(dt) => DatePattern::from_str_kind(s, kind, dt.to_owned()),
+            DatePatternContext::Fixed(dt) => DatePattern::from_str_kind(s, kind, dt.to_owned()),
```

Comment:
nit: `.to_owned()` is now redundant

Reply:
Thanks. I'm still trying to internalize the details of Rust ownership.

I thought I needed `.to_owned()` because `from_str_kind` takes a `DateTime<Tz>`, not `&DateTime<Tz>`, and Tz may not be Copy. But I think I don't anymore because `DateTime<Local>`, `DateTime<Utc>`, and `DateTime<FixedOffset>` are all Copy, so the compiler knows it can make a new owned value by just copying bits. Am I understanding correctly?

Reply:
They're all `Copy`, so `*self` (where: `self: &DatePatternContext`) can make a copy of `dt: &DateTime<_>`, and now `dt: DateTime<_>`. Another option is `*dt` (without `*self`.)